### PR TITLE
Remove quaternion package dependency

### PR DESCRIPTION
**Pull request by johnwlambert**
Switch to fully using [`scipy`](https://github.com/scipy/scipy) (8.3k stars) instead of [`quaternion`](https://github.com/moble/quaternion) (413 stars)

### CODE REVIEW

File: argoverse/evaluation/competition_util.py [hunk 1]
```diff
@@ -169,6 +169,7 @@ def dist(p1: Tuple[float, float], p2: Tuple[float, float]) -> float:
 
 
 def poly_to_label(poly: Polygon, category: str = "VEHICLE", track_id: str = "") -> ObjectLabelRecord:
+    """ """
```
Comment (benjaminrwilson) [inline argoverse/evaluation/competition_util.py:172]: Thanks for taking care of this, @johnwlambert. Could we add a short docstring here?

File: argoverse/utils/transform.py [hunk 2]
```diff
@@ -48,6 +55,13 @@ def quat_argo2scipy(q: np.ndarray) -> np.ndarray:
     return q_scipy
 
 
+def quat_scipy2argo(q: np.ndarray) -> np.ndarray:
```
Comment (benjaminrwilson) [inline argoverse/utils/transform.py:58]: This is on a bit of a meta note, but we really need to clean up `transform.py` (the file as a whole).

File: argoverse/evaluation/competition_util.py (hunk 1 above)
Comment (johnwlambert) [reply argoverse/evaluation/competition_util.py]: Sure -- just added one

**Comment by johnwlambert** [comment]
@tagarwal-argoai would you mind taking a look at this, also?

File: argoverse/evaluation/competition_util.py [hunk 3]
```diff
@@ -136,13 +134,9 @@ def get_polygon_from_points(points: np.ndarray) -> Polygon:
...
-    for simplex in hull.simplices:
-        poly.append([points[simplex, 0][0], points[simplex, 1][0], points[simplex, 2][0]])
-        poly.append([points[simplex, 0][1], points[simplex, 1][1], points[simplex, 2][1]])
-
-        # plt.plot(points[simplex, 0], points[simplex, 1], 'k-')
-
-    return Polygon(poly)
+    # `simplices` contains indices of points forming the simplical facets of the convex hull.
+    poly_pts = hull.points[np.unique(hull.simplices)]
+    return Polygon(poly_pts)
```
Comment (johnwlambert) [inline argoverse/evaluation/competition_util.py:139]: @janesjanes and @benjaminrwilson would you mind taking a brief look at this?

I believe the previous code was actually mixing the x,y,z coordinates of 3 vertices, which I don't think was the intended behavior.

If a simplicial facet is defined by vertex 0 = (x0,y0,z0), vertex 1=(x1,y1,z1), vertex 2=(x2,y2,z2), then I think we were forming (x0,x1,x2) as a point of the polygon, which wouldn't be correct.

File: argoverse/evaluation/competition_util.py [hunk 4]
```diff
@@ -158,57 +152,52 @@ def get_rotated_bbox_from_points(points: np.ndarray) -> Polygon:
     return get_polygon_from_points(points).minimum_rotated_rectangle
 
 
-def unit_vector(pt0: Tuple[float, float], pt1: Tuple[float, float]) -> Tuple[float, float]:
-    # returns an unit vector that points in the direction of pt0 to pt1
-    dis_0_to_1 = math.sqrt((pt0[0] - pt1[0]) ** 2 + (pt0[1] - pt1[1]) ** 2)
-    return (pt1[0] - pt0[0]) / dis_0_to_1, (pt1[1] - pt0[1]) / dis_0_to_1
-
+def poly_to_label(poly: Polygon, category: str = "VEHICLE", track_id: str = "") -> ObjectLabelRecord:
+    """Convert a Shapely Polygon to a 3d cuboid by estimating the minimum-bounding rectangle.
```
Comment (johnwlambert) [inline argoverse/evaluation/competition_util.py:156]: I tried to simplify this logic a bit here, and make things a bit clearer.

File: argoverse/evaluation/competition_util.py [hunk 5]
```diff
@@ -266,18 +255,14 @@ def save_label(argoverse_data: ArgoverseTrackingLoader, labels: List[ObjectLabel
...
             "center": {
                 "x": label.translation[0],
                 "y": label.translation[1],
                 "z": label.translation[2],
             },
-            "rotation": {
-                "x": label.quaternion[0],
-                "y": label.quaternion[1],
-                "z": label.quaternion[2],
-                "w": label.quaternion[3],
```
Comment (johnwlambert) [inline argoverse/evaluation/competition_util.py:279]: @janesjanes since we express quaternions as qw, qx, qy, qz, I'm not sure this is correct here.

Since the `quaternion` package uses qw, qx, qy, qz, but we are unpacking this as qx, qy, qz, qw.

File: argoverse/evaluation/competition_util.py [hunk 6]
```diff
@@ -158,57 +152,52 @@ def get_rotated_bbox_from_points(points: np.ndarray) -> Polygon:
...
 
-    x = bbox.exterior.xy[0]
-    y = bbox.exterior.xy[1]
     z = np.array([z for _, _, z in poly.exterior.coords])
+    height = max(z) - min(z)
 
-    # z = poly.exterior.xy[2]
-
-    d1 = dist((x[0], y[0]), (x[1], y[1]))
-    d2 = dist((x[1], y[1]), (x[2], y[2]))
```
Comment (johnwlambert) [inline argoverse/evaluation/competition_util.py:183]: i thought this indexing was a bit hard to follow, so i tried to simplify it a bit

File: argoverse/evaluation/competition_util.py (hunk 3 above)
Comment (benjaminrwilson) [reply argoverse/evaluation/competition_util.py:136]: Thanks for looking into this, @johnwlambert. Before I dive in, are we using this function anywhere?

File: argoverse/evaluation/competition_util.py (hunk 4 above)
Comment (benjaminrwilson) [reply argoverse/evaluation/competition_util.py:153]: Does this get called outside of this file?

File: argoverse/evaluation/competition_util.py (hunk 3 above)
Comment (johnwlambert) [reply argoverse/evaluation/competition_util.py:136]: Thanks Ben. So this is used in a tracking competition demo notebook: https://github.com/argoai/argoverse-api/blob/master/demo_usage/competition_tracking_tutorial.ipynb

File: argoverse/evaluation/competition_util.py (hunk 4 above)
Comment (johnwlambert) [reply argoverse/evaluation/competition_util.py:153]: Just here: https://github.com/argoai/argoverse-api/blob/master/demo_usage/competition_tracking_tutorial.ipynb

File: argoverse/evaluation/competition_util.py [hunk 7]
```diff
@@ -220,7 +206,7 @@ def poly_to_label(poly: Polygon, category: str = "VEHICLE", track_id: str = "")
 
 
 def get_objects(clustering: DBSCAN, pts: np.ndarray, category: str = "VEHICLE") -> List[Tuple[np.ndarray, uuid.UUID]]:
-
+    """ """
```
Comment (benjaminrwilson) [inline argoverse/evaluation/competition_util.py:209]: Could we provide a small docstring here?
